nemo_aligner/utils/distributed.py

```diff
@@ -191,12 +193,114 @@ def _compute_distributed_log_softmax(vocab_parallel_logits):
+            global_top_k_p_indices = torch.where(
+                sorted_indices_to_remove, torch.tensor(-1, dtype=torch.long), global_top_k_p_indices
+            )  # the top_p are kept, the rest are set to -1 (so will be filtered in the following lines)
+    else:
+        global_top_k_p_indices = torch.empty_like(
+            local_topk_indices, dtype=local_topk_indices.dtype, device=local_topk_indices.device
+        )  # [B,S,k]
+
+    torch.distributed.broadcast(global_top_k_p_indices, src=0)
+
+    # generate a mask according to the rank
+    # filter indices within the current rank's segment
+    mask_top_k_p_indices = (global_top_k_p_indices >= vocab_start_index) & (
+        global_top_k_p_indices < vocab_end_index
+    )  # [B,S,k] where only indices that are within the scope of current rank are True
+
+    # adjust indices to local index space
+    local_top_k_p_indices = global_top_k_p_indices - vocab_start_index
+    local_top_k_p_indices = torch.where(
+        mask_top_k_p_indices, local_top_k_p_indices, torch.tensor(-1, dtype=torch.long)
+    )  # [B,S,k] - the global top_k_p indices are localized to the rank, or -1 if they are not in this rank's segment
+
+    full_mask = torch.full_like(logits, -float("Inf"))
+    batch_indices, sequence_indices = torch.where(
+        mask_top_k_p_indices.any(dim=-1)
+    )  # collect all b,s indices where there is a valid index in [b,s,:]
+    local_vocab_indices = local_top_k_p_indices[
+        batch_indices, sequence_indices
+    ]  # collect the v indices per each [b,s]. should be up to k valid indices (not valid is -1)
+    # is there a better way to do it ?
+    valid_local_indx_mask = local_vocab_indices != -1
+    valid_local_batch_idx = batch_indices.unsqueeze(1).expand_as(valid_local_indx_mask)[valid_local_indx_mask]
+    valid_local_sequence_idx = sequence_indices.unsqueeze(1).expand_as(valid_local_indx_mask)[valid_local_indx_mask]
+    valid_local_vocab_idx = local_vocab_indices[valid_local_indx_mask]
+    full_mask[valid_local_batch_idx, valid_local_sequence_idx, valid_local_vocab_idx] = logits[
+        valid_local_batch_idx, valid_local_sequence_idx, valid_local_vocab_idx
+    ]
+    return full_mask
+
+
+def _distributed_apply_sampling_params(logits, context_lengths, sampling_params, rank, world_size):
+    # apply the sampling params to the logits - focusing only on the generated tokens.
+    if sampling_params.get("use_greedy", False):
+        return logits
+
+    context_length = context_lengths.min().item()
+    resp_logits = logits[:, context_length - 1 :]
+    # divide by temp
+    if sampling_params["temperature"] != 1.0:
+        resp_logits /= sampling_params["temperature"]
+    top_k = sampling_params["top_k"]
+    top_p = sampling_params["top_p"]
+    if top_k > 0:
+        # Note : currently assuming that top_p is applied only if top_k>0.
+        resp_logits = _compute_distributed_top_k_p(resp_logits, top_k, top_p, rank, world_size)
```

Comment:
Note that right now we do nothing with `resp_logits`. But let's address my comment regarding moving the context truncation optimization out of this function before fixing this.

Reply:
`resp_logits` is used as a view into a slice of `logits`. 
do you have a preferred way to write it ?